%% BioMed_Central_Tex_Template_v1.05
%%                                      %
%  bmc_article.tex            ver: 1.05 %
%                                       %


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <27 January 2006>           %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%									   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     


% \def\includegraphic{}
% \def\includegraphics{}



\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 

%Review style settings
% \newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% User-specific packages, macros, and settings
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
\usepackage{color} 
\usepackage{amsmath} 
% \usepackage{amsfonts} 
\usepackage{xspace}
\usepackage{bm} \bm{}
\usepackage{graphicx}
\usepackage{timestamp} 

\newcommand{\PMA}{\ensuremath{\textnormal{PM}_\textnormal{A}}\xspace}
\newcommand{\PMB}{\ensuremath{\textnormal{PM}_\textnormal{B}}\xspace}
\newcommand{\gAA}{\textnormal{AA}\xspace}
\newcommand{\gAB}{\textnormal{AB}\xspace}
\newcommand{\gBB}{\textnormal{BB}\xspace}
\newcommand{\eps}{\varepsilon\xspace}
\newcommand{\pkg}[1]{\textit{#1}\xspace}
\newcommand{\CN}{\textnormal{CN}\xspace}
\newcommand{\kb}{\textnormal{kb}\xspace}
\newcommand{\GWSSixf}{GenomeWideSNP\_6\xspace}
%\newcommand{\CP}{\ensuremath{\textnormal{CP}}\xspace}
%\newcommand{\CPs}{\ensuremath{\textnormal{CPs}}\xspace}
\newcommand{\CP}{\ensuremath{\textnormal{change point}}\xspace}
\newcommand{\CPs}{\ensuremath{\textnormal{change points}}\xspace}

\newcommand{\bx}{\mathbf{x}\xspace}
\newcommand{\by}{\mathbf{y}\xspace}
\newcommand{\bz}{\mathbf{z}\xspace}
\newcommand{\ba}{\mathbf{a}\xspace}
\newcommand{\bb}{\mathbf{b}\xspace}
\newcommand{\bA}{\mathbf{A}\xspace}
\newcommand{\bB}{\mathbf{B}\xspace}
\newcommand{\bC}{\mathbf{C}\xspace}
\newcommand{\bS}{\mathbf{S}\xspace}
\newcommand{\bU}{\mathbf{U}\xspace}
\newcommand{\bV}{\mathbf{V}\xspace}
\newcommand{\bW}{\mathbf{W}\xspace}
\newcommand{\beps}{\bm{\varepsilon}\xspace}
\newcommand{\btheta}{\bm{\theta}\xspace}
\newcommand{\bnu}{\bm{\nu}\xspace}

\usepackage{fancyhdr}

% Comment the 2nd row to highlight updates
% \newcommand{\updated}[2][red]{{\color{#1}{#2}}\xspace}
% \renewcommand{\updated}[2][red]{#2\xspace}

\graphicspath{{figures/col/}}

\newenvironment{TODO}{\color{red}\textbf{TODO:}}{}
\newenvironment{PN}{\color{blue}\textbf{PN:}}{}

\pagestyle{fancy}
\lhead{\sf TumorBoost}
\rhead{\sf \timestamp}


% Begin ...
\begin{document}
\begin{bmcformat}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{TumorBoost: A single-sample method for calibrating allele-specific tumor copy numbers in paired tumor-normal designs}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Ensure \and is entered between all but   %%
%% the last two authors. This will be       %%
%% replaced by a comma in the final article %%
%%                                          %%
%% Ensure there are no trailing spaces at   %% 
%% the ends of the lines                    %%     	
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\author{Henrik Bengtsson\correspondingauthor$^1$%
       \email{Henrik Bengtsson\correspondingauthor - hb@stat.berkeley.edu}%
      \and
         Pierre Neuvial$^1$%
         \email{Pierre Neuvial -pierre@stat.berkeley.edu}
       and 
         Terence P. Speed$^{1,2}$%
         \email{Terence P. Speed - terry@stat.berkeley.edu}%
   }
      

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address{%
    \iid(1) Department of Statistics, University of California, Berkeley, USA\\
    \iid(2) Bioinformatics Division, Walter \& Eliza Hall Institute of Medical Research, Parkville, Australia
}%

\maketitle

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% The Section headings here are those for  %%
%% a Research article submitted to a        %%
%% BMC-Series journal.                      %%  
%%                                          %%
%% If your article is not of this type,     %%
%% then refer to the Instructions for       %%
%% authors on http://www.biomedcentral.com  %%
%% and change the section headings          %%
%% accordingly.                             %%   
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\noindent { \sf This version: \timestamp}

\begin{abstract}
        % Do not use inserted blank lines (ie \\) until main body of text.
        \paragraph*{Background:} High-throughput genotyping microarrays can be used not only to assess changes in total DNA copy number but also changes in allele-specific copy numbers (ASCNs). Even after state of the art preprocessing methods, ASCN estimates for Affymetrix genotyping arrays still suffer from systematic effects that make them difficult to use effectively for downstream studies in cancers.
      
        \paragraph*{Results:} We propose a single-sample method for calibrating ASCN estimates of a tumor based on ASCN estimates of a paired normal. The method applies to any paired tumor-normal estimates regardless of technology and generation. We demonstrate that our method leads to a much clearer separation between different ASCN states, including \emph{copy number neutral events} that cannot be detected using total copy numbers only.

        \paragraph*{Conclusions:} Combined with single-array preprocessing methods, such as CRMA~v2, we conclude that ASCN estimates with high precisions can be obtained from a single pair of tumor-normal hybridizations, and recommend using paired tumor-normal DNA microrarray experiments when applicable.

        \paragraph*{Availability:}
A single-sample bounded-memory implementation is available in \pkg{aroma.cn}.
\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% The Section headings here are those for  %%
%% a Research article submitted to a        %%
%% BMC-Series journal.                      %%  
%%                                          %%
%% If your article is not of this type,     %%
%% then refer to the instructions for       %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and change the section headings          %%
%% accordingly.                             %% 
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

  \begin{PN}
Vocabulary:
    \begin{itemize}
    \item single ``sample'' or single ``individual'' method ?
    \item ``normal hybridization'', ``normal sample'', ``normal tissue'' ? Define it once and stick to it.
    \item ``allele B fraction'' / ``fraction of allele B''
 \end{itemize}
  \end{PN}

  \begin{TODO}
    \begin{itemize}
    \item test the method on Illumina data.
    \end{itemize}
  \end{TODO}


%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}
\label{secBackground}

The development of microarray technologies to assess DNA copy number changes over the past few years has been triggered by the fact that genomic alterations are hallmarks of gene deregulation and genome instability in cancers~\cite{albertson03chromosome,hanahan00the-hallmarks}. 

Recent technologies include genotyping microarrays that quantify not only total copy numbers (TCN) but also allele-specific copy numbers (ASCNs), that is, allelic contributions of each allele to TCN. ASCNs estimates are crucial as they can pinpoint genomic alterations that are TCN neutral, such as uniparental disomy (UPD), or allele-specific amplifications.

\begin{TODO}
  Technologies: Illumina and Affy.
\end{TODO}

Several preprocessing methods have been developped for these technologies~\cite{Affymetrix_2008m,BengtssonH_etal_2009b,BengtssonH_etal_2008,KornJ_etal_2008,LiWong_2001}.
\begin{TODO}
  REFs for Illumina preprocessing ?
\end{TODO}
The typical output of genotyping microarray preprocessing method for a given SNP unit is an estimation of the total copy number --- the sum of allele A and allele B intensities divided by a reference --- and the fraction $\beta$ of allele B intensity relative to the total intensity. Together, these two quantities determine the raw ASCNs at each SNP locus.

\subsection*{SNP effects after preprocessing}

The main motivation of this paper is that existing preprocessing methodsdo not fully correct for SNP-specific effects, that result in SNP-specific distributions of genotype clusters.
\begin{TODO}
  Mention, credit and discuss $^\star$RLMM and Bird$^\star$ approaches here.
\end{TODO}
 As a consequence, $\beta$ estimates are not well suited for comparisons across loci along the genome, as illustrated in Figure XXX. 

\begin{TODO}
Update these figures. Highlight a few points in both panels to illustrate the correspondence between the plots.
\end{TODO}

\resizebox{0.96\columnwidth}{!}{
 \includegraphics[height=5cm]{betaGenome}
  \includegraphics[height=5cm]{betaNVsBetaT}
}

%\caption{
\emph{Fraction of allele B $(\beta)$ in a tumor and its matched normal on chrosomosome XXX. Each point corresponds to a SNP. Left: $\beta$  along chromosome XXX for the tumor (top) and the normal (bottom). Right: Scatter plot of $\beta$ in the tumor vs $\beta$ in the normal. Black '+' indicate the expected location of normal genotype clusters: $(0,0)$, $(1/2,1/2)$ and $(1,1)$. Data were preprocessed using the CRMA~v2 method~\cite{BengtssonH_etal_2009b}.
}
%}

There is considerable deviation from the expected location of genotype clusters, even in the normal when no (or very few) genomic alterations are expected (Figure XXX, right).
\begin{TODO}
  make the text understandable without the figure caption.
\end{TODO}
Interestingly, this deviation is quite reproducible between the tumor and the normal: most points are scattered along the diagonal. As a consequence of this deviation, genomic profiles of allele B fraction show a high variability are  both in the tumor and the normal (Figure XXX, right). 

\begin{TODO}
  Add 2 other plots: $\beta, beta$ for a T/N pair with alterations in the tumor and a pair of unrelated samples.
\end{TODO}

\begin{TODO}
  add a second illustration to illustrate the motivation for downstream analyses, ie  $(\theta_A,\theta_B)$ for a given sample across SNPs in a region as in Nancy's talk:
\resizebox{0.96\columnwidth}{!}{\includegraphics{thetaAcrossSNPs}}
\end{TODO}

\subsection*{Proposed method}

In this paper we present a method calibrating the allele-specific copy-number estimates (ASCNs) of a tumor tissue given ASCNs of matched normal tissue or blood extract.  The method does not require external references and it is only the relative ASCNs that are calibrated: total CN estimates are neither used as an input nor adjusted.

We show that our method leads to a much clearer separation between different ASCN states, including \emph{copy number neutral events}. Such events cannot be detected using total copy numbers alone; however, detecting them is often crucial, for example in cancer studies where they can help finding new tumor suppressor genes.

\begin{PN}
  Next paragraph is pretty much repeated in the Evaluation section...
\end{PN}
Genotypes \emph{of the normal} hybridization are used for the calibration, hence the performance of our method depends on genotype quality. We compare the results obtained using genotypes inferred using a set of samples to those of a naive single-array genotyping. We argue that the proposed naive genotyping algorithm yields calibrated $\beta$ values that are good enough for typical downstream analyses such as the search of regions of Loss of Heterozygosity (LOH) or ASCN segmentation, as all these analyses  involve smoothing or segmentation of $\beta$ along the genome. The calibration method obtained using this genotyping algorithm is therefore a purely \emph{single-sample method}.


\subsection*{A single sample method}
The realization of a single-sample method has several implications:
(i) Each tumor-normal pair can be analyzed immediately without needing reference samples.
(ii) Samples can be processed in parallel on different hosts/processors making it possible to decrease the processing time of large data sets.
(iii) There is no need to reprocess a sample when new samples are produced, which further saves time and computational resources.
Furthermore, (iv) the decision to filter out poor samples can be made later, because a poor sample will not affect the processing of other samples.
More importantly, a single-sample method is
(v) more practical for applied medical diagnostics, because individual patients can be analyzed at once, even when they come singly rather than in batches.  This may otherwise be a limiting factor in projects with a larger number of samples.

\subsection*{Outline}
The outline of this paper is as follows. 
In Methods, we describe the underlying model, its estimation, and an interpretation in terms of allelic crosstalk. In Results, we show that the signal-to-noise ratios (SNRs) of the calibrated ASCNs are significantly larger than corresponding non-calibrated estimates.
In Discussion, we conclude the study, discuss potential limitations, extensions, and give future research directions.


 
%%%%%%%%%%%%%%%%%%
\section*{Methods}
Let $(\theta_{N,j,A}, \theta_{N,j,B})$ be the signal intensities after preprocessing for SNP $j$ in the normal ($N$) tissue, and $(\theta_{T,j,A}, \theta_{t,j,B})$ the corresponding intensitites in the tumor ($T$) tissue. The allele~B fraction~\cite{PeifferD_etal_2006} for SNP $j=1,\ldots,J$ in the normal is defined as:
\begin{eqnarray*}
  \beta_{N,j} = \frac{\theta_{N,j,A}}{\theta_{N,j}},
%   \label{eqnCnLogRatio}
\end{eqnarray*}
where $\theta_{N,j} = \theta_{N,j,A} + \theta_{N,j,B}$ is the non-polymorphic signal at locus $j$.
For a diploid SNP $j$, for which the genotype is either AA, AB or BB, we expect the allele~B fraction $\beta_{N,j}$ to be close to $0$, $1/2$, or $1$, respectively.
The allele~B fraction $\beta_{T,j}$ for the tumor is defined analoguously. Because we cannot expected a SNP in a tumor to be diploid at any random SNP, we cannot predict where $\beta_{T,j}$ falls. For a tumor-normal pair, we observe $(\beta_{N,j}, \beta_{T,j})$ at each SNP $j$.

\subsection*{Proposed model and its estimation}
Consider a SNP $j$ that is diploid in the normal tissue.  
Let the true genotype be denoted by the true allele~B fraction $\mu_{N,j}$ with possible states $\{0,1/2,1\}$ corresponding to genotypes $\{\gAA,\gAB,\gBB\}$. For the tumor, the true allele~B fraction is denoted by  $\mu_{T,j} \in [0,1]$. As the tumor needs not be diploid at SNP $j$, and the ``tumor sample'' can be contaminated by normal tissue, $\mu_{T,j}$ cannot be restricted \textit{a priori} to a discrete state space.

Based on this, for SNP $j$ we model the observed allele~B fraction for the tumor and the normal pair as
\begin{displaymath}
  \begin{cases}
    \beta_{T,j} & =  \mu_{T,j} + \delta_{j} + \eps_{T,j}\\
    \beta_{N,j}  & =  \mu_{N,j} + \delta_{j} + \eps_{N,j}
    % \label{eqnFracBModel}
  \end{cases}\,,
\end{displaymath}

where $\delta_{j}$ is a SNP-specific effect and $\eps_{T,j}$ and $\eps_{N,j}$ are independent zero-mean error terms. The main assumption of this model is that $\delta_{j}$ is the same for the tumor and the normal hybridization. It is motivated by the high correlation observed in Figure 1, right.

\begin{TODO}
  Explain what we do for X and Y chromosomes where the normal needs not be diploid.
\end{TODO}

The normal genotype $\mu_{N,j}$ is much easier to estimate than $\mu_{N,j}$ because it can be safely assumed to be either 0, 1/2, or 1. Assuming that such an estimate $\widehat{\mu_{N,j}}$ is available, $\delta_j$ can be estimated by  $\beta_{N,j} - \widehat{\mu_{N,j}}$ and a straightforward estimator of $\mu_{T,j}$ is 
\begin{displaymath}
     \widehat{\mu_{T,j}} = \beta_{T,j} + \widehat{\mu_{N,j}} - \beta_{N,j}\,.
\end{displaymath}

% The difference $\mu_{T,j}-\mu_{N,j}$ can then be estimated straightforwardly from the data.
% \begin{eqnarray*}
%   \widehat{\mu_{T,j}-\mu_{N,j}} = \beta_{T,j} - \beta_{N,j}\,.
% %   \label{eqnTumorBoostEstimate}
% \end{eqnarray*}

% \begin{TODO}
%   Illustration: plot of $\beta_N$ vs $\beta_T$ (with lines with $y-x=c$ for $c \in$ seq(-1, 1, length=5), say) and tracks of $\beta_T$, $\beta_N$ and $\beta_T-\beta_N$. 
% \end{TODO}

\begin{PN}
  This formulation of the model does not explicitly show that the pairing is essential. I think that the main assumption is that $\delta_j$ is \emph{normal-genotype-specific}, and that the model should first be formulated in a more generic (ie non necessariliy paired setting), say something like
\begin{displaymath}
   \beta_{H(i),j} =  \mu_{H(i),j} + \delta(\mu_{N(i), j}) + \eps_{H(i),j} \,,
\end{displaymath}
where $i$ is an individual, $H(i)$ is an hybridization, and $\delta(\mu_{N(i), j})$ is a normal-genotype-specific term for this SNP, which depends on the individual only through the normal genotype at this SNP. I like this formulation because it accounts for all what we observe in $(\beta, \beta)$ plots:

\begin{itemize}
\item in the paired case, the $\delta$s are equal because they concern the same individual at the same SNP, and the model reduces to the previous one;
\item in the unpaired case the $\delta$s don't cancel when the samples are not paired, unless normal genotype is the same, ie 
\begin{displaymath}
  \delta(\mu_{N(i), j}) = \delta(\mu_{N(i'), j})  \iff \mu_{N(i), j}=\mu_{N(i'), j}\,.
\end{displaymath}
This is also something that we see in $(\beta, \beta)$ plots for unpaired samples (high correlation along the diagonal, and no correlation elsewhere). 
\end{itemize}
Actually it does not really account for the observed \emph{anti-correlation} for SNPs which are say AA in sample 1 and BB in sample 2. I'm working on it ;)
\end{PN}

\subsection*{Model interpretation in terms of allelic crosstalk}
\label{secACCModel}
In~\cite{BengtssonH_etal_2008} and \cite{BengtssonH_etal_2009b} it is argued that there exist crosstalk between the allele signals.  For instance, for a diploid SNP that is truly AA we will not only observe a great signal in the \PMA probes, but also some signal in the \PMB probes.  One explanation for this is crosshybridization due to the close similarity of probe sequences.  In the afformentioned studies, the authors propose an offset and crosstalk correction that is applied globally (to the six different heterozygous groups).  In addition, \cite{BengtssonH_etal_2009b} suggest to apply an additional nucleotide-position probe sequence normalization, which further corrects for imbalances between the two alleles.  It is shown that these corrections significantly improve the ability to detect total CN changes.
However, when looking at the distribution of the allele-specific summaries for a particular SNP $j$ across a set of samples, that is $\{(\theta_{ijA},\theta_{ijA})\}_i$, it is clear that for some SNPs there exists a remaining crosstalk, which is likely to be SNP specific.  In Figure~XXX, the allele-specific summaries for SNP\_A-XXXXXX in data set XXXXXX are shown, which clearly shows that the two homozygous genotype groups AA and BB are not located along the axes, that is, they are not orthogonal. For this particular SNP the heterozygous group is located along the diagonal as expected.

\begin{TODO}
  Illustration: something like the following $(\theta_A,\theta_B)$ plots across samples for two or three consecutive SNPs (maybe with CNA, CNB instead to avoid the discussion about the copy number scale ?).
 Left: normal and tumor before calibration. Right: same with the calibrated tumor. Argue that the angle is really different from SNP to SNP. 
\resizebox{0.96\columnwidth}{!}{\includegraphics{thetaAcrossSamples} \includegraphics{thetaAcrossSamples}}
\end{TODO}

\begin{PN}
   I can draw a couple of such plots for a few units of interest.
\end{PN}
We propose the following crosstalk model for estimated allele-specific summaries $\{(\theta_{ijA},\theta_{ijA})\}_i$:
\begin{eqnarray*}
  \btheta_{ij} = \bS_i \bx_{ij} + \beps_{ij},
%   \label{eqnACCi}
\end{eqnarray*}
where the crosstalk matrix
\begin{eqnarray*}
 \bS_i = 
 \begin{bmatrix}
   S_{iAA} & S_{iAB} \\
   S_{iBA} & S_{iBB}
 \end{bmatrix}
%   \label{eqnACCii}
\end{eqnarray*}
is shared by all SNPs of sample $i$.

%--------------------------------------------------------------------------
% EVALUATION METHODS
%--------------------------------------------------------------------------
\subsection*{Evaluation methods}
\label{secEvaluation}
% The evaluation methods used in this paper are inspired by those used in \cite{BengtssonH_etal_2009b}. 

In order to assess the performance of TumorBoost, we compare the ability of the fraction of allele B to detect allele-specific genomic alterations before and after calibration. More specifically, we selected one tumor-normal pair for which we identified a region of copy number neutral LOH in the tumor, that is, a region in which the total copy number is 2 and there are homozygous and heterozygous loci in the normal, but there are only homozygous loci in the tumor. Importantly, such a region cannot be identified by analyzing total copy number only as it is not accompanied by any change in total copy number.

% Input: beta and genotype calls
 
\subsubsection*{Quantifying copy neutral LOH}
In order to identify a breakpoint between a normal region and a region of copy neutral LOH, SNPs that are homozygous in the normal are not informative because they will remain homozygous in the tumor. We therefore focus on SNPs that are \emph{heterozygous in the normal}, and calculate a transformed version of the fraction of allele B, denoted by $\rho$ and defined by

\begin{displaymath}
  \rho_j = \left\vert \beta_j - \frac{1}{2} \right\vert \,.
\end{displaymath}

This kind of transformation is widely used for downstream analyses~\cite{PeifferD_etal_2006,AttiyehE_etal_2009}
\begin{TODO}
  add other REFS ? probably all Illumina SNP papers.
\end{TODO}
 and can be motivated as follows for the particular case of a region of copy neutral LOH. In such a region, $\beta$ is expected to have only two bands, corresponding to the two homozygous states, whereas it has three bands in a normal region. As $\beta$ is symmetric around $1/2$, $\rho$ has only one band near 0 in a copy neutral LOH region, and two bands in a normal region. After excluding SNPs that are homozygous in the normal, $\rho$ also only has one band in a normal region, which is expected to be near 1/2. 
 \begin{PN}
   Painful explanation... draw a picture !
 \end{PN}

 \begin{PN}
   $<$sidetracking$>$
 \end{PN}
In order to interpret $\rho$ it is useful to note that it can be written as the minimum allele fraction:
\begin{displaymath}
  \rho = \beta \wedge (1-\beta)\,,
\end{displaymath}

or, equivalently, as the minimum ASCN, rescaled by the total copy number

\begin{displaymath}
  \rho = \frac{CN_A \wedge CN_B}{CN_A + CN_B}\,.
\end{displaymath}
\begin{PN}
  ASCNs and TCN have not been formally defined yet.  $<$/sidetracking$>$
\end{PN}

Data points in the  XXXkb region centered around the breakpoint were excluded. The remaining data points are annotated to belong to either the normal state or the copy neutral LOH state. We use a  Receiver Operator Characteristics (ROC) analysis to assess how well the fraction of allele B separates between the neutral and the copy neutral LOH data points.

\subsubsection*{Resolution}
\begin{TODO}
  see if this section is relevant...
\end{TODO}
This evaluation is done on the full-resolution $\beta$ as well as on a smoothed 
version, where $\beta$s are smoothed by using non-overlapping bins 
for which the average $\beta$ is calculated. This approach is motivated by the fact that downstream analyses of ASCN use smoothing or segmentation and are therefore concerned with the influence of the amount of smoothing on the output of the analysis. It was inspired 
by total copy number studies~\cite{LaiW_etal_2005,WillenbrockFridlyand_2005}. A similar approach has been used more recently in~\cite{BengtssonH_etal_2009b,BengtssonH_etal_2008}

\subsubsection*{Sensitivity of the calibration to genotype calls}
Genotypes \emph{of the normal} are used for the calibration, hence the performance of our method depends on genotype quality. To evaluate sensitivity of the calibration to genotype calls we compare the results obtained using genotypes inferred using a set of samples to those of a naive single-array genotyping which we now describe.

For a given normal hybridization, we estimate the density of $\beta$ using a kernel density estimator.
\begin{PN}
  more details ?
\end{PN}
We define genotype classes by thresholding $\beta$ at local minima of this density estimate. There local minima are determined using a discrete derivative (ie successive differences) of the density estimate. 
\begin{PN}
Is this done for each chromosome separately ? What about sexual chromosomes ? 
\end{PN}

We argue that the proposed naive genotyping algorithm yields calibrated $\beta$ values that are good enough for typical downstream analyses such as the search of regions of Loss of Heterozygosity (LOH) or ASCN segmentation, as all these analyses  involve smoothing or segmentation of $\beta$ along the genome. The calibration method obtained using this genotyping algorithm is therefore a purely \emph{single-sample method}.

 
As mentioned above we only our evaluation method focuses on SNPs that are \emph{heterozygous in the normal sample}. Therefore the results of this evaluation will depend on the genotype calling method used for the normal sample. To address this point we compare results with naive genotyping, TGCA genotypes, and ``random'' genotypes.

\subsubsection*{Sensitivity of the evaluation to genotype calls}
As mentioned above we only our evaluation method focuses on SNPs that are \emph{heterozygous in the normal sample}. Therefore the results of this evaluation will depend on the genotype calling method used for the normal sample. To address this point we compare results with naive genotyping, TGCA genotypes, and ``random'' genotypes.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Results and Discussion %%
%%
\section*{Results}
\label{secResults}

%--------------------------------------------------------------------------
% DATA SET
%--------------------------------------------------------------------------
\subsection*{Data set}
We used data from the  Cancer Genome Atlas (TCGA) project~\cite{CollinsBarker_2007,TCGA_2008c}, a collaborative initiative to better understand cancer using existing large-scale whole-genome technologies.  Several tumor types are or are planned to be studied, including brain cancer (glioblastoma multiforme; GBM), ovarian cancer and lung cancer. 
 
From the Data Coordinating Center (\url{http://tcga-data.nci.nih.gov/tcga/homepage.htm}), we downloaded (Jan 2009) raw data (Level~1; CEL files) for a set of GBM tumor-normal pairs.
For the purpose of illustrating our method, we will focus mainly on sample TCGA-02-0620
(vials 01A vs 10A), \begin{TODO}
  check vials
\end{TODO}
 because it has a region of copy neutral LOH Chr~17.

\subsection*{Allele~B fraction before and after calibration}

\begin{PN}
  Add $(\theta_A, \theta_B)$ plots for each sample in a right panel (colored by region ?) and/or $(\widetilde{\beta_T}, \beta_N)$ plots to each of the next 3 plots in order to get another sense of the improvement ?
\end{PN}

% \begin{figure}[!tpbh]
\begin{center}
\resizebox{0.96\columnwidth}{!}{\includegraphics{TCGA-02-0069,tumor,Chr05,fracB,beforeAndAfter}}%
 \label{figROCs,chr05}
\end{center}
%  \caption{
%  }
% 
% \end{figure}

% \begin{figure}[!tpbh]
\begin{center}
  \resizebox{0.96\columnwidth}{!}{\includegraphics{TCGA-02-0069,tumor,Chr06,fracB,beforeAndAfter}}%
\label{figROCs,chr06}
\end{center}
%  \caption{
%  }
% \end{figure}


% \begin{figure}[!tpbh]
\begin{center}
  \resizebox{0.96\columnwidth}{!}{\includegraphics{TCGA-02-0069,tumor,Chr18,fracB,beforeAndAfter}}%
 \label{figROCs,chr18}
\end{center}
%  \caption{
%  }
% \end{figure}


\subsection*{Detecting copy number neutral LOH}
We use one pair of tumor/normal hybridizations to assess how well ASCN estimates (or $\rho$ estimates) can differentiate between normal and copy neutral LOH states. Given a global threshold $\tau$, a locus with $\rho$ below $\tau$ is considered to belong to the copy neutral LOH state, otherwise the normal state. By calculating the fraction of correctly called copy neutral LOH loci, we obtain an estimate of 
the true positive rate, and by calculating the fraction of incorrectly called normal loci, we obtain an estimate of the false positive rate. By adjusting the threshold, we can estimate a ROC curve. 

\subsubsection*{Full resolution $\beta$}
The true positive rate of calling a copy neutral LOH (among normal loci) as a function of false-positive rate is depicted in Figure XXX. The ROC curves show that\dots

\subsubsection*{Smoothed $\beta$}
\begin{TODO}
  See if this section is relevant.
\end{TODO}
Using a windowing technique similar to that in \cite{BengtssonH_etal_2009b,BengtssonH_etal_2008a} for a fixed false-positive rate we can estimate the true positive rate as a function of amount of smoothing. Since a given amount of smoothing corresponds to a given distance between loci this provides us with a first approximation to the effective resolution of a method. In the upper panel of Figure XXX the true-positive rate (for normal v. copy neutral LOH) as a function of resolution is shown before and after calibration.

\section*{Discussion}

\subsection*{Downstream analysis methods}
= increasing power to detect copy number events (even non copy neutral), and making it possible to detect copy number neutral events from Affymetrix genotyping arrays, e.g.: 
\begin{itemize}
\item Segmentation of ASCN when normal genotypes are available
\item Iterative segmentation of ASCN when normal genotypes are not available
\item Quantile segmentation of ASCN when normal genotypes are not available
\end{itemize}

 When genotype calls for the normal sample are available they can be used to segment tumor ASCNs using any copy number segmentation algorithm. When these genotype calls are not avaiable we propose two algorithms that achieve joint segmentation of normal and tumor ASCNs, both leading to genotype calls for the normal sample as a by-product.\\

\subsection*{Influence of genotyping errors on calibration}
In the Results section we have shown that our calibration methods leads to an improved signal ratio at the chromosome or at the genome scale. However the correction factor is genotype-specific, so if the normal genotype call for a given SNP is wrong, the correction factor will actually add bias to the estimated fraction of allele B. We argue that this is not a serious issue, for two main reasons.

First, we can take genotype \emph{confidence scores} into account, either by discarding SNPs for which we are not confident in the genotype or by  using the confidence scores in whatever downstream analyses. 
\begin{TODO}
  in suppl. mat.: a section with beta along the genome after calibration for different levels of stringency for genotype calls (for the naive genotyping)
\end{TODO}

Second, although our method does improve allele-specific copy number calls at the single locus level whenever normal genotype calls are correct, our goal is to improve the SNR of the fraction of allele B \emph{along the genome} in order to facilitate downstream analyses. In the results section we showed that our calibration method leads to a significant improvement of this SNR regardless of the genotype calling algorithm chosen. 

\subsection*{Extensions of the method}
\begin{itemize}
\item applicability / usefulness for Illumina data. paragraph on why Affy is more noisy than Illumina ? Sequences are all the same for Illumina, hence no probe sequence specific effects.
\item multi source version
\item multi sample (unpaired) version ?
\end{itemize}

    

%%%%%%%%%%%%%%%%%%%%%%
\section*{Conclusions}
  Text for this section \ldots


\section*{Algorithm and implementation}
The TumorBoost calibration method is available in R~\cite{RDevel_2008} package \pkg{aroma.cn} part of the \pkg{aroma.affymetrix} framework~\cite{BengtssonH_etal_2008b}.  The  method is designed and implemented to have bounded-memory usage, regardless of the number of samples/arrays processed. 
Furthermore, the complexity of the algorithm is linear in the number of loci ($J$).
Since it is a single-sample method, the tumor-normal pairs can be calibrated in parallel on multiple hosts/processors.
The method applies to estimates obtained by any SNP technology.




    
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Authors contributions}
    Text for this section \ldots

    

%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}

We gratefully acknowledge the TCGA Consortium and all its members for the TCGA Project initiative, for providing samples, tissues, data processing, and making data and results available.

\emph{Funding}: NCI grant U24 CA126551.

\emph{Conflict of interest}: none declared.

 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%              
%%  Bmc_article.bst  will be used to                       %%
%%  create a .BBL file for submission, which includes      %%
%%  XML structured for BMC.                                %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %% 
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %% 
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}

%  \bibliographystyle{bmc_article}  % Style BST file
%   \bibliography{bmc_article} }     % Bibliography file (usually '*.bib' ) 
\bibliographystyle{bmc_article}
\bibliography{bioinformatics-journals-abbr,hb-at-maths.lth.se,pnRefs}}
% \bibliography{hb-at-maths.lth.se}

%%%%%%%%%%%

\ifthenelse{\boolean{publ}}{\end{multicols}}{}



\end{bmcformat}

\end{document}




