\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt,a4paper]{article}    

% Load packages
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
\usepackage{fancyvrb}

% To refer to labels to the manuscript
\usepackage{xr} 
\externaldocument[CRMAv2:]{BengtssonH_2008c-CRMAv2}
 
 
\setlength{\topmargin}{-2.0cm}
\setlength{\textheight}{25.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% Controlling page disposition of Tables and Figures
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
\renewcommand{\topfraction}{1.0}	% max fraction of floats at top
\renewcommand{\bottomfraction}{1.0}	% max fraction of floats at bottom
\renewcommand{\textfraction}{0.07}
 
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% User-specific packages
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
\usepackage{amsmath} 
\usepackage{xspace}
\usepackage{bm} \bm{}
\usepackage{graphicx}


% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% Graphics settings
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% The search path (within curly brackets and separated by commas)
% where to find graphics files
\graphicspath{{figures/col/},{figures/}}
% Make dvips deal with gzip'ed EPS figures.
\DeclareGraphicsRule{.eps.gz}{eps}{.eps.bb}{`gunzip -c #1}


% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% User-specific macros
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
\newcommand{\GWSFive}{GWS5\xspace}
\newcommand{\GWSSix}{GWS6\xspace}
\newcommand{\GWSFivef}{GenomeWideSNP\_5\xspace}
\newcommand{\GWSSixf}{GenomeWideSNP\_6\xspace}

\newcommand{\PMA}{\ensuremath{\textnormal{PM}_\textnormal{A}}\xspace}
\newcommand{\PMB}{\ensuremath{\textnormal{PM}_\textnormal{B}}\xspace}
\newcommand{\TP}{\ensuremath{\textnormal{TP}}\xspace}
% MathOperator!
\newcommand{\MAD}{\ensuremath{\textnormal{MAD}}\xspace}

\newcommand{\chrX}{ChrX\xspace}
\newcommand{\chrY}{ChrY\xspace}
\newcommand{\NspI}{\emph{Nsp}I\xspace}
\newcommand{\StyI}{\emph{Sty}I\xspace}
\newcommand{\Nsp}{\ensuremath{\textnormal{Nsp}}\xspace}
\newcommand{\Sty}{\ensuremath{\textnormal{Sty}}\xspace}

\newcommand{\filename}[1]{\textit{#1}\xspace}
%\newcommand{\url}[1]{\href{#1}{#1}\xspace}
\newcommand{\pkg}[1]{\textit{#1}\xspace}
\newcommand{\args}[1]{\textit{#1}\xspace}
\newcommand{\FL}{\textnormal{FL}\xspace}
\newcommand{\GC}{\textnormal{GC}\xspace}
\newcommand{\CN}{CN\xspace}
\newcommand{\kb}{\textnormal{kb}\xspace}
\newcommand{\PM}{\textnormal{PM}\xspace}
\newcommand{\MM}{\textnormal{MM}\xspace}
\newcommand{\bx}{\mathbf{x}\xspace}
\newcommand{\by}{\mathbf{y}\xspace}
\newcommand{\bz}{\mathbf{z}\xspace}
\newcommand{\ba}{\mathbf{a}\xspace}
\newcommand{\bA}{\mathbf{A}\xspace}
\newcommand{\bB}{\mathbf{B}\xspace}
\newcommand{\bC}{\mathbf{C}\xspace}
\newcommand{\bS}{\mathbf{S}\xspace}
\newcommand{\bU}{\mathbf{U}\xspace}
\newcommand{\bV}{\mathbf{V}\xspace}
\newcommand{\bW}{\mathbf{W}\xspace}
\newcommand{\beps}{\bm{\varepsilon}\xspace}
\newcommand{\bnu}{\bm{\nu}\xspace}
\renewcommand{\Re}{\mathbb{R}\xspace}
\newcommand{\II}{\mathbb{I}\xspace}
\DeclareMathOperator{\median}{\textnormal{median}}
\DeclareMathOperator{\mad}{\textnormal{mad}}

\newcommand{\updated}[3][red]{{{\color{#1}\textsl{\textbf{#2}}}\endnote{#3 \textsl{#2}}}\xspace}
\renewcommand{\updated}[3][red]{#2\xspace} 

\usepackage{natbib}
%\newcommand{\citet}[1]{\cite{#1}} 
%\newcommand{\citep}[1]{\cite{#1}} 


% Begin ...
\begin{document}


\title{Supplementary Notes \#1:\\Summary of annotation data files for the Affymetrix GenomeWideSNP\_5 and GenomeWideSNP\_6 chip types}
\author{Henrik Bengtsson et al.}
\maketitle
\tableofcontents

\clearpage
\section{Annotation data files}
Although the features on the arrays never change, their annotations might get updated as the Human Genome databases get updated.  In the CRMA~v2 study, we used the latest annotation data available on the Affymetrix website as of August 2008.
We will list all annotation files needed by CRMA~v2 for analyzing data for the \GWSFivef (GWS5) and the \GWSSixf (GWS6) chip types.  First, we specify all relevant files available from Affymetrix (Tables~\ref{tblGWSFiveAffymetrix}~\&~\ref{tblGWSSixAffymetrix}).  Then we specify the files compiled from the Affymetrix files for use in aroma.affymetrix (Tables~\ref{tblGWSFiveAromaAffymetrix}~\&~\ref{tblGWSSixAromaAffymetrix}).

\subsection{Affymetrix annotation data}

\begin{table}[hp]
\begin{center}
\begin{tabular}{lp{0.7\textwidth}}
\hline
File name    & GenomeWideSNP\_5,r2.cdf \\
File size    & 239,670,727 bytes \\
MD5 checksum & e062cb42a392dd74721d5b7cf9a286f1 \\
Notes        & Affymetrix default chip definition file (CDF). \\
\hline
File name    & GenomeWideSNP\_5,Full,r2.cdf \\
File size    & 261,578,689 bytes \\
MD5 checksum & 79f7a8353b4978dedbeff05a7897ff6e \\
Notes        & Affymetrix full chip definition file (CDF). \\
\hline
File name    & GenomeWideSNP\_5.CN\_probe\_tab \\
File size    & 43,252,969 bytes \\
MD5 checksum & e661544a27163d1242ca690a486cb6b5 \\
Notes        & Affymetrix file probe sequence file for CN probes. \\
\hline
File name    & GenomeWideSNP\_5.probe\_tab \\
File size    & 233,706,497 bytes \\
MD5 checksum & 69b66720591fde9333acf5eb4d1b3e68 \\
Notes        & Affymetrix file probe sequence file for SNPs. \\
\hline
File name    & GenomeWideSNP\_5.cn.na26.annot.csv \\
File size    & 180,849,309 bytes \\
MD5 checksum & ec4f6cb4b482923d73d07f1b07faefe4 \\
Notes        & Affymetrix NetAffx v26 annotation file for CN probes. \\
\hline
File name    & GenomeWideSNP\_5.na26.annot.csv \\
File size    & 755,337,946 bytes \\
MD5 checksum & af59235b6fccada7f871257149a89215 \\
Notes        & Affymetrix NetAffx v26 annotation file for SNPs. \\
\hline
\end{tabular}
\end{center}
\caption{Details of all Affymetrix specific GenomeWideSNP\_5 annotation files used in the study.}
\label{tblGWSFiveAffymetrix}
\end{table}


\begin{table}[hp]
\begin{center}
\begin{tabular}{lp{0.7\textwidth}}
\hline
File name    & GenomeWideSNP\_6.cdf \\
File size    & 484,489,553 bytes \\
MD5 checksum & 223f3cd9141404b2a926a40cf47d6f1a \\
Notes        & Affymetrix default chip definition file (CDF). \\
\hline
File name    & GenomeWideSNP\_6.Full.cdf \\
File size    & 493,291,745 bytes \\
MD5 checksum & 3fbe0f6e7c8a346105238a3f3d10d4ec \\
Notes        & Affymetrix full chip definition file (CDF). \\
\hline
File name    & GenomeWideSNP\_6.CN\_probe\_tab \\
File size    & 96,968,290 bytes \\
MD5 checksum & 3dc2d3178f5eafdbea9c8b6eca88a89c \\
Notes        & Affymetrix file probe sequence file for CN probes. \\
\hline
File name    & GenomeWideSNP\_6.probe\_tab \\
File size    & 341,479,928 bytes \\
MD5 checksum & 2037c033c09fd8f7c06bd042a77aef15 \\
Notes        & Affymetrix file probe sequence file for SNPs. \\
\hline
File name    & GenomeWideSNP\_6.cn.na26.annot.csv \\
File size    & 482,222,873 bytes \\
MD5 checksum & 948eb406774aa5097590debd0d667a22 \\
Notes        & Affymetrix NetAffx v26 annotation file for CN probes. \\
\hline
File name    & GenomeWideSNP\_6.na26.annot.csv \\
File size    & 1,628,608,540 bytes \\
MD5 checksum & 323f9afa0c180c146260b5eb689d0bd2 \\
Notes        & Affymetrix NetAffx v26 annotation file for SNPs. \\
\hline
\end{tabular}
\end{center}
\caption{Details of all Affymetrix specific GenomeWideSNP\_6 annotation files used in the study.}
\label{tblGWSSixAffymetrix}
\end{table}


\clearpage
\subsection{Imported annotation data}
Several of Affymetrix annotation files use file formats that are not intended to be used directly in a computational system, but rather be imported once.  The \pkg{aroma.affymetrix} framework utilizes its own binary files that are more compact and faster to access.  The contents of these are imported from the above Affymetrix annotation data files.  Further details on the source files used for each compiled file is given in the file footer of each file.  To download these and for further details, see the aroma.affymetrix webpage.

\begin{table}[hp]
\begin{center}
\begin{tabular}{lp{0.7\textwidth}}
\hline
File name    & GenomeWideSNP\_5,HB20080710.acs \\
File size    & 121,981,027 bytes \\
MD5 checksum & bd0da64b09ea164082e066798774a3c5 \\
Notes        & Aroma.affymetrix probe sequence data from above Affymetrix sequence files. \\
\hline
File name    & GenomeWideSNP\_5,Full,r2,na26,HB20080822.ufl \\
File size    & 3,684,511 bytes \\
MD5 checksum & a533ac3ba64f36902d13bed8fe2a9a5b \\
Notes        & Aroma.affymetrix fragment length data mapping to the full CDF.  Compiled from above Affymetrix NetAffx files. \\
\hline
File name    & GenomeWideSNP\_5,Full,r2,na26,HB20080822.ugp \\
File size    & 4,605,439 bytes \\
MD5 checksum & 9827c9fad08144d6e590b87984350a26 \\
Notes        & Aroma.affymetrix genome location data mapping to the full CDF.  Compiled from above Affymetrix NetAffx files. \\
\hline
File name    & GenomeWideSNP\_5,r2,na26,HB20080822.ufl \\
File size    & 3,445,230 bytes \\
MD5 checksum & 566812ef309486baf0a8496bb8a35ef5 \\
Notes        & Aroma.affymetrix fragment length data mapping to the default CDF.  Compiled from above Affymetrix NetAffx files. \\
\hline
File name    & GenomeWideSNP\_5,r2,na26,HB20080822.ugp \\
File size    & 4,306,339 bytes \\
MD5 checksum & 9330fdcb30b3a3b4c13cbdeb8de4ffb5 \\
Notes        & Aroma.affymetrix genome location data mapping to the default CDF.  Compiled from above Affymetrix NetAffx files. \\
\hline
\end{tabular}
\end{center}
\caption{Details of all aroma.affymetrix specific GenomeWideSNP\_5 annotation files used in the study.}
\label{tblGWSFiveAromaAffymetrix}
\end{table}



\begin{table}[hp]
\begin{center}
\begin{tabular}{lp{0.7\textwidth}}
\hline
File name    & GenomeWideSNP\_6,HB20080710.acs \\
File size    & 179,217,531 bytes \\
MD5 checksum & f04f081e0a1900653d957a8f320744c0 \\
Notes        & Aroma.affymetrix probe sequence data from above Affymetrix sequence files. \\
\hline
File name    & GenomeWideSNP\_6,Full,na26,HB20080722.ufl \\
File size    & 7,526,454 bytes \\
MD5 checksum & 6f11e9bd3a7a0cb060d5fcf671b0776a \\
Notes        & Aroma.affymetrix fragment length data mapping to the full CDF.  Compiled from above Affymetrix NetAffx files. \\
\hline
File name    & GenomeWideSNP\_6,Full,na26,HB20080821.ugp \\
File size    & 9,407,937 bytes \\
MD5 checksum & 5a7bef30a458cb238ae2167aa41f5bd6 \\
Notes        & Aroma.affymetrix genome location data mapping to the full CDF.  Compiled from above Affymetrix NetAffx files. \\
\hline
File name    & GenomeWideSNP\_6,na26,HB20080821.ufl \\
File size    & 7,425,058 bytes \\
MD5 checksum & 522b89d875f39832f5423e78cffba8c8 \\
Notes        & Aroma.affymetrix fragment length data mapping to the full CDF.  Compiled from above Affymetrix NetAffx files. \\
\hline
File name    & GenomeWideSNP\_6,na26,HB20080821.ugp \\
File size    & 9,281,127 bytes \\
MD5 checksum & ad63ef009b44f1274f6e2a35cb951dbc \\
Notes        & Aroma.affymetrix genome location data mapping to the default CDF.  Compiled from above Affymetrix NetAffx files. \\
\hline
\end{tabular}
\end{center}
\caption{Details of all aroma.affymetrix specific GenomeWideSNP\_6 annotation files used in the study.}
\label{tblGWSSixAromaAffymetrix}
\end{table}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% CDF files
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\clearpage
\section{Summary of annotation data}

This section provides a summary of the relevant annotation data available for the GWS5 and the GWS6 chip types.  We distinguish between the annotation data available in the Chip Definition Files (CDFs) and the NetAffx files, because the former are less likely to change over time whereas the latter gets updates when the genome annotations get updated.  Affymetrix' NetAffx annotation data are updated several times a year.

\subsection{Affymetrix CDF data}

Affymetrix provides one ``default'' and one ``full'' CDF for each of the GWS5 and GWS6 chip types.  See Table~\ref{tblGWSSixAffymetrix} for these files.
The default CDF contains a subset of the full CDF where certain SNP units have been filtered out due to homology to other regions and poor performance (private communication with Affymetrix).
Tables~\ref{tblCdfUnits}~\&~\ref{tblCdfProbes} provides a summary of these CDFs.
All summaries presented here and in the main paper refer to the full CDFs, unless stated otherwise.

\subsubsection{Unit annotations}

\begin{table}[hp]
\begin{center}
\begin{tabular}{|l|rr||rr|}
\hline
                        &  \GWSFive &  \GWSFive &   \GWSSix  &   \GWSSix \\
                        &  default  &  full        & default &   full    \\
\hline
\hline
UNIT TYPES              &           &           &            &           \\
CN units                &  417,269  &  417,269  &   945,826  &   945,826 \\
SNPs                    &  440,794  &  500,568  &   906,600  &   931,946 \\
Subtotal                &  858,063  &  917,847  & 1,852,426  & 1,877,772 \\
AFFX-SNPs               &    3,022  &    3,022  &     3,022  &     3,022 \\
Subtotal                &  861,085  &  920,859  & 1,855,448  & 1,880,794 \\
Others                  &       24  &       69  &       621  &       621 \\
Total                   &  861,109  &  920,928  & 1,856,069  & 1,881,415 \\
\hline									            						             
EXCLUDED FROM FULL      &           &           &            &           \\
CN units                &        0  &        -  &         0  &         - \\
SNPs                    &   59,744  &        -  &    25,346  &         - \\
AFFX-SNPs               &        0  &        -  &         0  &         - \\
Others                  &       45  &        -  &         0  &         - \\
\hline									            						             
CN UNIT STRANDNESS 		  &	  				&	  				&	 				   &	 		     \\
Sense only              &        0  &        0  &         0  &         0 \\
Antisense only          &  417,269  &  417,269  &   945,826  &   945,826 \\
Opposite strands        &        0  &        0  &         0  &         0 \\
Both strands            &        0  &        0  &         0  &         0 \\
\hline									            						             
SNP STRANDNESS					&	  				&	  				&	 				   &	 		     \\
Sense only              &  234,449  &  260,266  &   477,538 &    491,830 \\
Antisense only          &  174,549  &  194,126  &   429,062  &   440,116 \\
Opposite strands        &   31,796  &   46,176  &         0  &         0 \\
Both strands            &        0  &        0  &         0  &         0 \\
\hline									            						             
AFFX-SNP STRANDNESS 		&	  				&	  				&	 				   &	 		     \\
Sense only              &      145  &      145  &       145  &       145 \\
Antisense only          &      129  &      129  &       129  &       129 \\
Opposite strands        &        0  &        0  &         0  &         0 \\
Both strands            &    2,748  &    2,748  &     2,748  &     2,748 \\
\hline									            						             
SNP ALIGNMENT	           &	  			 &           &	 		      &	 		      \\
Aligned allele pairs     &  285,984  &  308,169  &   906,600  &   931,946 \\
Non-aligned allele pairs &  154,810  &  192,399  &         0  &         0 \\
\hline									            						             
AFFX-SNP ALIGNMENT	     &	  			 &	  				&	 		       &	 		      \\
Aligned allele pairs     &    3,022  &    3,022  &    3,022  &     3,022  \\
Non-aligned allele pairs &        0  &        0  &        0  &         0  \\
\hline									            						             
\end{tabular}
\end{center}
\caption{Summary of the unit annotation available in the CDF files.  All values are in counts.}  %% Summary generated 2008-08-22.
\label{tblCdfUnits}
\end{table}


\clearpage
\subsubsection{Probe annotations}
\begin{table}[hp]
\begin{center}
\begin{tabular}{|l|rr||rr|}
\hline
                        &  \GWSFive &  \GWSFive &   \GWSSix  &   \GWSSix \\
                        &  default  &  full        & default &   full    \\
\hline
\hline

PROBES                  &           &           &            &           \\
CN probes               &   417,269 &   417,269 &    945,826 &   945,826 \\
SNP probes              & 3,526,352 & 4,004,544 &  5,660,710 & 5,833,210 \\
Subtotal                & 3,943,621 & 4,421,813 &  6,606,536 & 6,779,036 \\
AFFX-SNP probes         &    81,504 &    81,504 &     81,504 &    81,504 \\
Subtotal                & 4,025,125 & 4,503,317 &  6,688,040 & 6,860,540 \\
Other probes            &   178,055 &   188,239 &     32,420 &    32,420 \\
Excluded from full      &   488,376 &         0 &    172,500 &         0 \\
Total                   & 4,691,556 & 4,691,556 &  6,892,960 & 6,892,960 \\
\hline									            						             
EXCLUDED FROM FULL      &           &           &            &           \\
CN probes               &        0  &        -  &         0  &         - \\
SNP probes              &  478,192  &        -  &   172,500  &         - \\
AFFX-SNP probes         &        0  &        -  &         0  &         - \\
Other probes            &   10,184  &        -  &         0  &         - \\
\hline									            						             
CN UNIT PROBES           &	  	     &	  			 &	 				 &	 				 \\
CN units with 1 probe    & 417,269  &  417,269  &   945,826 &   945,826 \\
\hline
SNP PROBE PAIRS         &	  				&	  				&	 				   &	 				 \\
SNPs with 3 pairs       &        0  &        0  &    796,045 &   811,179 \\
SNPs with 4 pairs       &  440,794  &  500,568  &    110,555 &   120,767 \\
\hline
AFFX-SNP PROBE PAIRS     &	         &	  			 &	 				  &	 		      \\
SNPs with 12 pairs       &    2,461  &    2,461  &     2,461  &     2,461 \\
SNPs with 20 pairs       &      561  &      561  &       561  &       561 \\
\hline
\end{tabular}
\end{center}
\caption{Summary on the probe annotation available in the CDF files.  All values are in counts.}    %% Summary generated 2008-08-22.
\label{tblCdfProbes}
\end{table}



\subsection{Affymetrix NetAffx data}

Affymetrix makes so called NetAffx CSV files available for download.  These files contain annotation data exported from their NetAffx data base.  For each chip type there exists one or more NetAffx CSV files, e.g. for the GWS chip types there is one for the CN units and one for all other units on the chip.  See Tables~\ref{tblGWSFiveAffymetrix}~\&~\ref{tblGWSSixAffymetrix} for which these files are.
The NetAffx data base is updated frequently and the following information is likely to get slightly outdated over time. 

\subsubsection{Genome positions}
In Table~\ref{tblGenomicLocation} we summarize how many SNPs and CN loci have known annotations according to NetAffx.  
The genome location per chromosome is summarized in Table~\ref{tblGenomicLocationByChromosome} for each of the default and the full CDF.

We do not know why the location is unknown for some loci, but from our investigation we believe it is mainly because such loci map to multiple positions in the genome.  For instance, in NetAffx release 26, there is no genomic location reported for SNP\_A-4228947 (on GWS5, GWS6 and Mapping250K\_Nsp), and according to the NCBI SNP data base it (rs11261805) maps to the two locations 41,240,208 and 43,493,496 on Chr9.

Furthermore, for GWS5 the NetAffx annotation files currently available only contain data on the units in the default CDF.  This is confirmed by comparing the names of the units with known locations in the default and the full GWS5.

\begin{table}[hp]
\begin{center}
\begin{tabular}{|l||rr||rr|}
\hline
     & \multicolumn{2}{c||}{\GWSFive} & \multicolumn{2}{c|}{\GWSSix} \\
     & default & full* & default & full \\
     & \#loci & \#loci & \#loci & \#loci \\
\hline
\hline
SNPs with known locations        & 440,094 &  440,094  &   905,386 &   929,967 \\
SNPs with unknown locations      &     700 &   60,474  &     1,214 &     1,979 \\
\hline
CN probes with known locations   & 312,384 &  312,384  &   945,806 &   945,806 \\
CN probes with unknown locations & 104,885 &  104,885  &        20 &        20 \\
\hline
AFFX-SNPs with known locations   &   3,012 &    3,012  &     3,012 &     3,012 \\
AFFX-SNPs with unknown locations &      10 &       10  &        10 &        10 \\
\hline
Total with known locations       & \textbf{755,490} & \textbf{755,490} & \textbf{1,854,204}  & \textbf{1,878,785} \\
Total with unknown locations     & 165,369 &   92,121  &    1,244  &     2,009 \\
\hline
Total                            & 861,085 &  920,859  & 1,855,448 & 1,880,794 \\
\hline
\end{tabular}
\end{center}
\caption{Summary of genomic location data that is available in the NetAffx files of contents in GWS5 and GWS6 with respect to unit and probe class and availability of annotation data.  The effective sets of units available for CN analysis are emphasized in bold.  *See text for why the default and the full GWS5 are identical.  NetAffx v26 was used for this summary.}
\label{tblGenomicLocation}
\end{table}

\begin{table}[hp]
\begin{center}
\begin{tabular}{|r|r||rr||rr|}
\hline
     & & \multicolumn{2}{c||}{\GWSFive} & \multicolumn{2}{c|}{\GWSSix} \\
     & & default & full* & default & full \\
chromosome & seq. length (Mbs) & \#loci & \#loci & \#loci & \#loci \\
\hline
\hline
1  & 245.2   & 58,548  &  58,548  &  144,499  &  146,401  \\
2  & 243.3   & 63,380  &  63,380  &  151,902  &  153,663  \\
3  & 199.4   & 53,120  &  53,120  &  126,337  &  127,766  \\
4  & 191.6   & 50,594  &  50,594  &  118,933  &  120,296  \\
5  & 181.0   & 48,661  &  48,661  &  114,333  &  115,672  \\
6  & 170.7   & 47,329  &  47,329  &  111,440  &  112,825  \\
7  & 158.4   & 39,325  &  39,325  &   99,818  &  100,996  \\
8  & 145.9   & 41,134  &  41,134  &   97,040  &   98,277  \\
9  & 134.5   & 31,563  &  31,563  &   81,036  &   82,168  \\
10 & 135.5   & 39,140  &  39,140  &   92,331  &   93,592  \\
\hline
11 & 135.0   & 37,948  &  37,948  &   88,295  &   89,525  \\
12 & 133.5   & 36,422  &  36,422  &   86,209  &   87,321  \\
13 & 114.2   & 28,119  &  28,119  &   65,310  &   66,067  \\
14 & 105.3   & 23,621  &  23,621  &   56,339  &   57,103  \\
15 & 100.1   & 20,968  &  20,968  &   52,810  &   53,556  \\
16 &  90.0   & 20,718  &  20,718  &   53,329  &   54,182  \\
17 &  81.7   & 17,411  &  17,411  &   46,024  &   46,632  \\
18 &  77.8   & 21,870  &  21,870  &   51,510  &   52,093  \\
19 &  63.8   & 10,631  &  10,631  &   29,855  &   30,299  \\
20 &  63.6   & 17,911  &  17,911  &   43,052  &   43,628  \\
\hline
21 &  47.0   & 10,412  &  10,412  &   24,787  &   25,111  \\
22 &  49.5   &  9,346  &   9,346  &   24,000  &   24,484  \\
X  & 152.6   & 26,373  &  26,373  &   86,064  &   87,198  \\
Y  &  51.0   &    946  &     946  &    8,841  &    9,485  \\
\hline
Mitochondrial  &  16.6kb &      -  &       -  &      110  &      445  \\
\hline
total  &     & 755,490 &  755,490 & 1,854,204 & 1,878,785 \\
\hline
\end{tabular}
\end{center}
\caption{Distribution of loci by chromosome for the different CDFs of GWS5 and GWS6.  *See text for why the default and the full GWS5 are identical.  NetAffx v26 was used for this summary.}
\label{tblGenomicLocationByChromosome}
%% Chromosome (sequence) lengths: http://www.ornl.gov/sci/techresources/Human_Genome/posters/chromosome/faqs.shtml
%% Summaries generated 2008-08-22 using GenomeWideSNP_5,r2,na26,HB20080822 and GenomeWideSNP_5,Full,r2,na26,HB20080822.
\end{table}


\clearpage
\subsubsection{Restriction enzymes and PCR fragment lengths}
\begin{table}[hp]
\begin{center}
\begin{tabular}{|l||rr||rr|}
\hline
     & \multicolumn{2}{c||}{\GWSFive} & \multicolumn{2}{c|}{\GWSSix} \\
     & default & full* & default & full \\
     & \#loci & \#loci & \#loci & \#loci \\
\hline
\hline
SNPs on NspI only        & 116,979 &  116,979  &   240,001 &   246,080 \\
SNPs on StyI only        &  74,135 &   74,135  &   154,884 &   160,899 \\
SNPs on both             & 248,980 &  248,980  &   510,330 &   522,472 \\
\hline
SNPs with known lengths  & 440,094 &  440,094  &   905,215 &   929,451 \\
SNPs with unknown lengths&     700 &   60,474  &     1,385 &     2,495 \\
\hline
\hline
CN probes on NspI only   & 140,099 &  140,099  &   451,191 &   451,191 \\
CN probes on StyI only   &   1,208 &    1,208  &         0 &         0 \\
CN probes on both        & 171,077 &  171,077  &   494,615 &   494,615 \\
\hline
CN probes with known lengths   & 312,384 &  312,384  &   945,806 &   945,806 \\
CN probes with unknown lengths & 104,885 &  104,885  &        20 &        20 \\
\hline
\hline
AFFX-SNPs with known lengths   &       0 &        0  &         0 &         0 \\
AFFX-SNPs with unknown lengths &   3,022 &    3,022  &     3,022 &     3,022 \\
\hline
\hline
Total with known lengths       & 752,478 &  752,478  & 1,851,021 & 1,875,257 \\
Total with unknown lengths     & 108,607 &  168,381  &     4,427 &     5,537 \\
\hline
Total                          & 861,085 &  920,859  & 1,855,448 & 1,880,794 \\
\hline
\end{tabular}
\end{center}
\caption{Summary of fragment-length data that is available in the NetAffx files of contents in GWS5 and GWS6 with respect to unit and probe class and availability of annotation data.  *See text for why the default and the full GWS5 are identical.  NetAffx v26 was used for this summary.}
\label{tblFragmentLengths}
\end{table}





\clearpage
\section{Multi-enzyme digestion}
\label{secMultiEnzymeDigestion}
For the 100K as well as the 500K SNP-only assays, DNA is prepared in two parallel processes, each digesting the DNA using a unique restriction enzyme, amplifying the fragments by PCR, and hybridizing the products to separate arrays.  In the GWS assays, which like 500K use enzymes \NspI and \StyI, the two mixes of PCR products are no longer hybridized to separate arrays but instead hybridized in aliquot to the same array~\citep{Affymetrix_2007f,Affymetrix_2007g}.
Consequently, SNP target DNA of PCR products originating from different digestions may hybridize to the same probe, which is something that has to be taken into account when, for instance, fitting the fragment-length normalization.
For CN probes the situation is somewhat different.  Affymetrix selected the CN probes from a large pool of CN probes based on their performance on copy numbers (private communication).  This pilot study was conducted on a specially designed in-house chip set containing probes that are known to be on an \NspI fragment.  For this reason, some of the selected probes are exclusively on \NspI fragments, some are by chance both on \NspI and \StyI fragments, but none are exclusively on \StyI fragments.  Note, when annotation for the human genome is updated, some of the probes might by chance be re-annotated to become \StyI-only probes.
We have found that it is important that the preprocessing models these differences, otherwise there is a substantial risk for getting systematic biases between SNPs and CN probes due to enzymatic mixing imbalances.
%% [We will later see that the hybridization of two PCR products on the same array requires a special model in order to control for confounded effects due to mixing inbalances.]
See Table~\ref{tblFragmentLengths} for details on fragment-length information for the two chip types and the two enzymes.

%% Note that there is no enzyme annotation data available for a large set of the GWS5 SNP units.  Instead, we use \GWSSix annotation data for the common SNPs and for CN probes for which annotation data otherwise was missing.  For further details and updates on this problem, see The aroma.affymetrix website [\url{http://www.braju.com/R/aroma.affymetrix/}].


\clearpage
\bibliography{bioinformatics-journals-abbr,hb-at-maths.lth.se}
\bibliographystyle{natbib}
 

\end{document}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% HISTORY:
% 2009-04-09
% o Made this Supplementary Notes standalone.  It contains the summary
%   of the annotation data and nothing else.  It may be distributed
%   for other purposes unrelated to CRMA v2.
% o Created.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
