\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[11pt,a4paper]{article}    

% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
\usepackage{fancyvrb}
 
 
\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% User-specific packages
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
\usepackage{amsmath} 
\usepackage{xspace}
\usepackage{bm} \bm{}
\usepackage{graphicx}


% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% Graphics settings
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% The search path (within curly brackets and separated by commas)
% where to find graphics files
\graphicspath{{figures/col/},{figures/}}
% Make dvips deal with gzip'ed EPS figures.
\DeclareGraphicsRule{.eps.gz}{eps}{.eps.bb}{`gunzip -c #1}


% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
% User-specific macros
% - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
\newcommand{\GWSFive}{GWS5\xspace}
\newcommand{\GWSSix}{GWS6\xspace}
\newcommand{\GWSFivef}{GenomeWideSNP\_5\xspace}
\newcommand{\GWSSixf}{GenomeWideSNP\_6\xspace}

\newcommand{\PMA}{\ensuremath{\textnormal{PM}_\textnormal{A}}\xspace}
\newcommand{\PMB}{\ensuremath{\textnormal{PM}_\textnormal{B}}\xspace}
\newcommand{\TP}{\ensuremath{\textnormal{TP}}\xspace}
% MathOperator!
\newcommand{\MAD}{\ensuremath{\textnormal{MAD}}\xspace}

\newcommand{\chrX}{ChrX\xspace}
\newcommand{\chrY}{ChrY\xspace}
\newcommand{\NspI}{\emph{Nsp}I\xspace}
\newcommand{\StyI}{\emph{Sty}I\xspace}
\newcommand{\Nsp}{\ensuremath{\textnormal{Nsp}}\xspace}
\newcommand{\Sty}{\ensuremath{\textnormal{Sty}}\xspace}

\newcommand{\filename}[1]{\textit{#1}\xspace}
%\newcommand{\url}[1]{\href{#1}{#1}\xspace}
\newcommand{\pkg}[1]{\textit{#1}\xspace}
\newcommand{\args}[1]{\textit{#1}\xspace}
\newcommand{\FL}{\textnormal{FL}\xspace}
\newcommand{\GC}{\textnormal{GC}\xspace}
\newcommand{\CN}{CN\xspace}
\newcommand{\kb}{\textnormal{kb}\xspace}
\newcommand{\PM}{\textnormal{PM}\xspace}
\newcommand{\MM}{\textnormal{MM}\xspace}
\newcommand{\bx}{\mathbf{x}\xspace}
\newcommand{\by}{\mathbf{y}\xspace}
\newcommand{\bz}{\mathbf{z}\xspace}
\newcommand{\ba}{\mathbf{a}\xspace}
\newcommand{\bA}{\mathbf{A}\xspace}
\newcommand{\bB}{\mathbf{B}\xspace}
\newcommand{\bC}{\mathbf{C}\xspace}
\newcommand{\bS}{\mathbf{S}\xspace}
\newcommand{\bU}{\mathbf{U}\xspace}
\newcommand{\bV}{\mathbf{V}\xspace}
\newcommand{\bW}{\mathbf{W}\xspace}
\newcommand{\beps}{\bm{\varepsilon}\xspace}
\newcommand{\bnu}{\bm{\nu}\xspace}
\renewcommand{\Re}{\mathbb{R}\xspace}
\newcommand{\II}{\mathbb{I}\xspace}
\DeclareMathOperator{\median}{\textnormal{median}}
\DeclareMathOperator{\mad}{\textnormal{mad}}

\newcommand{\updated}[3][red]{{{\color{#1}\textsl{\textbf{#2}}}\endnote{#3 \textsl{#2}}}\xspace}
\renewcommand{\updated}[3][red]{#2\xspace} 

\newcommand{\citet}[1]{\cite{#1}} 
\newcommand{\citep}[1]{\cite{#1}} 


% Begin ...
\begin{document}


\title{Supplementary materials for CRMA v2}
\maketitle


\section*{Annotation data}
Although the features on the arrays never change, their annotations might get updated as the Human Genome databases get updated.  In this study, we have used the latest annotation data available on the Affymetrix website as of August 2008.  Table~\ref{tblAnnotationDataFiles} gives full details on the data files used.

File name: GenomeWideSNP\_5.r2.cdf \\
File size: 239,670,727 bytes \\
MD5 checksum: e062cb42a392dd74721d5b7cf9a286f1 \\
Notes: Affymetrix \\

File name: GenomeWideSNP\_5.Full.r2.cdf \\
File size: 261,578,689 bytes \\
MD5 checksum: 79f7a8353b4978dedbeff05a7897ff6e \\
Notes: Affymetrix \\

File name: GenomeWideSNP\_6.cdf \\
File size: 484,489,553 bytes \\
MD5 checksum: 223f3cd9141404b2a926a40cf47d6f1a \\
Notes: Affymetrix \\

File name: GenomeWideSNP\_6.Full.cdf \\
File size: 493,291,745 bytes \\
MD5 checksum: 3fbe0f6e7c8a346105238a3f3d10d4ec \\
Notes: Affymetrix \\

File name: GenomeWideSNP\_6.CN\_probe\_tab \\
File size:  \\
MD5 checksum: \\
Notes: Affymetrix, NetAffx~r26 \\

File name: GenomeWideSNP\_6.probe\_tab \\
File size: \\
MD5 checksum: \\ 
Notes: Affymetrix, NetAffx~r26 \\



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% CDF files
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{CDF annotation data}

Affymetrix provides various Chip Definition Files (CDFs) for both the \GWSFivef and the \GWSSixf chip type.
For \GWSSix there exist one ``full'' CDF (GenomeWideSNP\_6.Full.cdf) and one ``default'' CDF (GenomeWideSNP\_6.cdf).  The latter is a subset of the former where SNP units have been filter out due to homology to other regions and poor performance (private communication with Affymetrix).
For \GWSFive the corresponding CDFs are GenomeWideSNP\_5.Full.r2.cdf and GenomeWideSNP\_5.Full.r2.cdf, respectively.
Tables~S\ref{tblCdfUnits}~\&~S\ref{tblCdfProbes} provides a summary of these CDFs.
All summaries presented here and in the main paper refer to the full CDFs, if otherwise not stated.


\begin{table}
\begin{center}
\begin{tabular}{lrrrr}
                        &  \GWSFive &  \GWSFive &   \GWSSix  &   \GWSSix \\
                        &  default  &  full        & default &   full    \\
\hline									            						             
\hline									            						             
UNIT TYPES              &           &           &            &           \\
CN units                &  417,269  &  417,269  &   945,826  &   945,826 \\
SNPs                    &  440,794  &  500,568  &   906,600  &   931,946 \\
Subtotal                &  858,063  &  917,847  & 1,852,426  & 1,877,772 \\
AFFX-SNPs               &    3,022  &    3,022  &     3,022  &     3,022 \\
Subtotal                &  861,085  &  920,859  & 1,855,448  & 1,880,794 \\
Others                  &       24  &       69  &       621  &       621 \\
Total                   &  861,109  &  920,928  & 1,856,069  & 1,881,415 \\
\hline									            						             
EXCLUDED FROM FULL      &           &           &            &           \\
CN units                &        0  &        -  &         0  &         - \\
SNPs                    &   59,744  &        -  &    25,346  &         - \\
AFFX-SNPs               &        0  &        -  &         0  &         - \\
Others                  &       45  &        -  &         0  &         - \\
\hline									            						             
CN UNIT STRANDNESS 		  &	  				&	  				&	 				   &	 		     \\
Sense only              &        0  &        0  &         0  &         0 \\
Antisense only          &  417,269  &  417,269  &   945,826  &   945,826 \\
Opposite strands        &        0  &        0  &         0  &         0 \\
Both strands            &        0  &        0  &         0  &         0 \\
\hline									            						             
SNP STRANDNESS					&	  				&	  				&	 				   &	 		     \\
Sense only              &  234,449  &  260,266  &   477,538 &    491,830 \\
Antisense only          &  174,549  &  194,126  &   429,062  &   440,116 \\
Opposite strands        &   31,796  &   46,176  &         0  &         0 \\
Both strands            &        0  &        0  &         0  &         0 \\
\hline									            						             
AFFX-SNP STRANDNESS 		&	  				&	  				&	 				   &	 		     \\
Sense only              &      145  &      145  &       145  &       145 \\
Antisense only          &      129  &      129  &       129  &       129 \\
Opposite strands        &        0  &        0  &         0  &         0 \\
Both strands            &    2,748  &    2,748  &     2,748  &     2,748 \\
\hline									            						             
SNP ALIGNMENT	           &	  			 &           &	 		      &	 		      \\
Aligned allele pairs     &  285,984  &  308,169  &   906,600  &   931,946 \\
Non-aligned allele pairs &  154,810  &  192,399  &         0  &         0 \\
\hline									            						             
AFFX-SNP ALIGNMENT	     &	  			 &	  				&	 		       &	 		      \\
Aligned allele pairs     &    3,022  &    3,022  &    3,022  &     3,022  \\
Non-aligned allele pairs &        0  &        0  &        0  &         0  \\
\hline									            						             
\end{tabular}
\end{center}
\caption{Summary on the unit annotation available in the CDF files.  All values are in counts. [2008-08-22]}
\label{tblCdfUnits}
\end{table}

\begin{table}
\begin{center}
\begin{tabular}{lrrrr}
                        &  \GWSFive &  \GWSFive &   \GWSSix  &   \GWSSix \\
                        &  default  &  full        & default &   full    \\
\hline									            						             
\hline									            						             
PROBES                  &           &           &            &           \\
CN units                &   417,269 &   417,269 &    945,826 &   945,826 \\
SNPs                    & 3,526,352 & 4,004,544 &  5,660,710 & 5,833,210 \\
Subtotal                & 3,943,621 & 4,421,813 &  6,606,536 & 6,779,036 \\
AFFX-SNPs               &    81,504 &    81,504 &     81,504 &    81,504 \\
Subtotal                & 4,025,125 & 4,503,317 &  6,688,040 & 6,860,540 \\
Others                  &   178,055 &   188,239 &     32,420 &    32,420 \\
Excluded from full      &   488,376 &         0 &    172,500 &         0 \\
Total                   & 4,691,556 & 4,691,556 & 6,892,960  & 6,892,960 \\
\hline									            						             
EXCLUDED FROM FULL      &           &           &            &           \\
CN probes               &        0  &        -  &         0  &         - \\
SNPs                    &  478,192  &        -  &    25,346  &         - \\
AFFX-SNPs               &        0  &        -  &         0  &         - \\
Others                  &   10,184  &        -  &         0  &         - \\
\hline									            						             
CN UNIT PROBES           &	  	     &	  			 &	 				 &	 				 \\
CN units with 1 probe    &  417,269  &  417,269  &   945,826 &   945,826 \\
\hline
SNP PROBE PAIRS         &	  				&	  				&	 				   &	 				 \\
SNPs with 3 pairs       &        0  &        0  &    796,045 &   811,179 \\
SNPs with 4 pairs       &  440,794  &  500,568  &    110,555 &   120,767 \\
\hline
AFFX-SNP PROBE PAIRS     &	         &	  			 &	 				  &	 		      \\
SNPs with 12 pairs       &    2,461  &    2,461  &     2,461  &     2,461 \\
SNPs with 20 pairs       &      561  &      561  &       561  &       561 \\
\hline
\end{tabular}
\end{center}
\caption{Summary on the probe annotation available in the CDF files.  All values are in counts. [2008-08-22]}
\label{tblCdfProbes}
\end{table}



\subsubsection*{NetAffx annotation data}


\subsubsection*{Genome positions}

\begin{table}
\begin{center}
\begin{tabular}{lrr}
                        &  \GWSFive &   \GWSSix \\
                        &  full     &   full    \\
\hline
GENOME POSITION  			         	 &	  				&	 				  \\
SNPs with known locations        &  495,071  &   930,662 \\
SNPs with unknown locations      &    5,497  &     1,284 \\
CN probes with known locations   &  330,645  &   945,806 \\
CN probes with unknown locations &   86,624  &        20 \\
Total with known locations       & \textbf{825,716}  & \textbf{1,876,468} \\
Total with unknown locations     &   92,121  &     1,304 \\
\hline
\end{tabular}
\end{center}
\caption{Summary of genomic location data that is available in the NetAffx files of contents in \GWSFive and \GWSSix with respect to unit and probe class and availability of annotation data.  The effective sets of units available for CN analysis are emphasized in bold.}
\label{tblGenomicLocation}
\end{table}

\clearpage
\begin{Verbatim}[fontsize=\small]
GenomeWideSNP_5,r2,na26,HB20080822: GenomeWideSNP_5,Full,r2,na26,HB20080822:
        min       max     n                 min       max     n
1      5130 247188467 58548         1      5130 247188467 58548
2      6289 242717062 63380         2      6289 242717062 63380
3     47843 199380415 53120         3     47843 199380415 53120
4       507 191250146 50594         4       507 191250146 50594
5     72569 180668700 48661         5     72569 180668700 48661
6    102372 170804144 47329         6    102372 170804144 47329
7     52912 158811019 39325         7     52912 158811019 39325
8     21194 146264218 41134         8     21194 146264218 41134
9     40464 140185941 31563         9     40464 140185941 31563
10    67341 135355061 39140         10    67341 135355061 39140
11   188951 134449982 37948         11   188951 134449982 37948
12    20704 132287718 36422         12    20704 132287718 36422
13 17960319 114116756 28119         13 17960319 114116756 28119
14 18085724 106356482 23621         14 18085724 106356482 23621
15 18278389 100303907 20968         15 18278389 100303907 20968
16    18196  88799382 20718         16    18196  88799382 20718
17     6888  78638431 17411         17     6888  78638431 17411
18    33310  76115554 21870         18    33310  76115554 21870
19    40794  63782986 10631         19    40794  63782986 10631
20     9307  62426743 17911         20     9307  62426743 17911
21  9758743  46942557 10412         21  9758743  46942557 10412
22 14494864  49583993  9346         22 14494864  49583993  9346
23   142664 154582820 26373         23   142664 154582820 26373
24  2725328  57443119   946         24  2725328  57443119   946
\end{Verbatim}
%% [2008-08-22]			 


\clearpage
\begin{Verbatim}[fontsize=\small]
GenomeWideSNP_6,na26,HB20080821:    GenomeWideSNP_6,Full,na26,HB20080821:
        min       max      n                min       max      n
1     51599 247191012 144499        1     51599 247191012 146401
2      2785 242738130 151902        2      2785 242738130 153663
3     35346 199380516 126337        3     35346 199380516 127766
4      2282 191261905 118933        4      2282 191261905 120296
5     68533 180722927 114333        5     68533 180722927 115672
6     94662 170892931 111440        6     94662 170892931 112825
7     52912 158819766  99818        7     52912 158819766 100996
8     21255 146268960  97040        8     21255 146268960  98277
9     36587 140211216  81036        9     36431 140211216  82168
10    62760 135356695  92331        10    62760 135356695  93592
11   188510 134449982  88295        11   188510 134449982  89525
12    20704 132288263  86209        12    20704 132288263  87321
13 17924950 114126500  65310        13 17924950 114126500  66067
14 18072125 106356482  56339        14 18072125 106356482  57103
15 18276342 100286564  52810        15 18276342 100286564  53556
16      778  88815037  53329        16      778  88815037  54182
17      527  78643088  46024        17      527  78643088  46632
18     1543  76116029  51510        18     1543  76116029  52093
19    41911  63789667  29855        19    41911  63789667  30299
20     9306  62426598  43052        20     9306  62426598  43628
21  9758743  46921386  24787        21  9758743  46921386  25111
22 14432529  49581322  24000        22 14432529  49581322  24484
23   108478 154887040  86064        23   108478 154907376  87198
24   169542  57427648   8841        24   169542  57442197   9485
25      410     16149    110        25      195     16521    445
\end{Verbatim}
%% [2008-08-22]			 



\subsubsection*{Restriction enzymes and PCR fragment lengths}


\begin{Verbatim}[fontsize=\small]
GenomeWideSNP_5,r2,na26,HB20080822:
                snp    cnp affxSnp other  total
enzyme1-only 116979 140099       0     0 257078
enzyme2-only  74135   1208       0     0  75343
both         248980 171077       0     0 420057
missing         700 104885    3022    24 108631
total        440794 417269    3022    24 861109

GenomeWideSNP_5,Full,r2,na26,HB20080822:
                snp    cnp affxSnp other  total
enzyme1-only 116979 140099       0     0 257078
enzyme2-only  74135   1208       0     0  75343
both         248980 171077       0     0 420057
missing       60474 104885    3022    69 168450
total        500568 417269    3022    69 920928

GenomeWideSNP_6,na26,HB20080821:
                snp    cnp affxSnp other   total
enzyme1-only 240001 451191       0     0  691192
enzyme2-only 154884      0       0     0  154884
both         510330 494615       0     0 1004945
missing        1385     20    3022   621    5048
total        906600 945826    3022   621 1856069

GenomeWideSNP_6,Full,na26,HB20080722:
                snp    cnp affxSnp other   total
enzyme1-only 246080 451191       0     0  697271
enzyme2-only 160899      0       0     0  160899
both         522472 494615       0     0 1017087
missing        2495     20    3022   621    6158
total        931946 945826    3022   621 1881415
\end{Verbatim}
%% [2008-08-22]


\begin{table}
\begin{center}
\begin{tabular}{lrr}
                        &  \GWSFive &   \GWSSix \\
                        &  full     &   full    \\
\hline
FRAGMENT-LENGTH INFORMATION    	&	  				&	 				  \\
SNPs with known lengths         &  494,245  &   930,559 \\
SNPs with unknown lengths       &    6,323  &     1,387 \\
CN probes with known lengths    &  330,645  &   945,826 \\
CN probes with unknown lengths  &   86,624  &         0 \\
Total with known lengths        &  824,890  & 1,876,385 \\
Total with unknown lengths      &   92,947  &     1,387 \\
\hline
SNPS CUT BY EACH ENZYME      & 	  			&	 				  \\
SNPs only on NspI            &  129,953  &   246,340 \\
SNPs only on StyI            &   90,418  &   161,064 \\
SNPs on both                 &  273,874  &   523,155 \\
SNPs with no annotation      &    6,323\parbox{0mm}{$^*$}  &     1,387\parbox{0mm}{$^*$} \\
Total                        &  500,568  &   931,946 \\
\hline
CN PROBES CUT BY EACH ENZYME & 	  			&	 				  \\
CN probes only on NspI       &  149,548  &   451,191 \\
CN probes only on StyI       &        0  &         0 \\
CN probes on both            &  181,097  &   494,615 \\
CN probes with no annotation &   86,624\parbox{0mm}{$^*$}  &         0 \\
Total                        &  417,269  &   945,826 \\
\hline
\end{tabular}
\end{center}
\caption{Summary of annotation data related to restriction enzyme available in the NetAffx files of contents in \GWSFive and \GWSSix with respect to unit and probe class and availability of annotation data.  The same SNPs and CN probes that lack information on digestion enzyme also lack information on genome position.}
\label{tblEnzymeData}
\end{table}




\section*{Multi-enzyme digestion}
\label{secMultiEnzymeDigestion}
For the 100K as well as the 500K SNP-only assays, DNA is prepared in two parallel processes, each digesting the DNA using a unique enzyme, amplifying the fragments by PCR, and hybridizing the products to separate arrays.  In the GWS assays, which similarly to 500K uses enzymes \NspI and \StyI, the two mixes of PCR products are no longer hybridized to separate arrays but instead hybridized in aliqout to the same array~\citep{Affymetrix_2007f,Affymetrix_2007g}.
Consequently, SNP target DNA of PCR products originating from different enzymes may hybridize to the same probe, which is something that has to be taken into account when, for instance, fitting the fragment-length normalization.
For CN probes the situation is somewhat different.  Affymetrix selected the CN probes from a large pool of CN probes based on their performance on copy numbers (private communication).  This pilot study was conducted on a specially designed in-house chip set containing probes that are known to be on an \NspI fragment.  For this reason, some of the selected probes are exclusively on \NspI fragments, some are by chance both on \NspI and \StyI fragments, but none are exclusively on \StyI fragments.  Note, when annotation for the human genome is updated, some of the probes might by chance be reannotated to become \StyI-only probes.
We have found that it is important that the preprocessing models these differences, otherwise there is a substantial risk for getting systematic biases between SNPs and CN probes due to enzymatic mixing imbalances.
%% [We will later see that the hybridization of two PCR products on the same array requires a special model in order to control for confounded effects due to mixing inbalances.]
See Table~\ref{tblSNP5SNP6} and Table~\ref{tblSNP5SNP6enzymes} for details on fragment-length information for the two chip types and the two enzymes.

Note that there is no enzyme annotation data available for a large set of the \GWSFive SNP units.  Instead, we use \GWSSix annotation data for the common SNPs and for CN probes for which annotation data otherwise was missing.  For further details and updates on this problem, see The aroma.affymetrix website [\url{http://www.braju.com/R/aroma.affymetrix/}].





%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% OTHER MODELS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{How raw copy numbers were estimated by other models}
\label{secOtherMethods}
In addition to CRMA~v2, two external methods were evaluated in this paper.  The first is Affymetrix' \emph{CN5} method~\citep{Affymetrix_2008m}, and the second is implemented in the dChip software~\citep{LiWong_2001}.

\subsection*{CN5}
The CN5 method is implemented in the 'apt-copynumber-workflow' software part of the Affymetrix Power Tools (APT) v1.10.0.  The Affymetrix Genotyping Console (GTC)~v3.0 software \citep{Affymetrix_2008m} utilizes APT for CN5 estimates.  We choose to run GTC, because it is not fully documented what settings should be used for APT.  According to Affymetrix both approaches produce identical results (Affymetrix Scientific Community Forums, Thread: 'copy number: Genotyping Console 3.0 vs. apt 1.10.0?' on August 15, 2008).
In CN5, probe signals are normalized ('adapter-type background correction') for systematic variation due to so called \emph{enzyme recognition-sequence class}.   Next, all probe signals (excluding control probes) are quantile normalized using the Affymetrix 'sketch' algorithm.  For SNPs, chip effects $\{(\theta_{Aij}, \theta_{Bij})\}$ (as in the log-additive model of RMA) are estimated separately  for the two alleles using the plier algorithm.  The total CNs are obtained as in Eqns~\eqref{eqnSumSnpAB}-\eqref{eqnCnAB}.
Log ratios are calculated as in Eqn~\eqref{eqnLogRatio}, where the reference is $\theta_{Rj}=\median_i\{\theta_{ij}\}$ with the important difference that for ChrX (ChrY) it is only samples that empirically are found to females (males) that are included.  Finally, the raw CNs (log-ratios) are shifted such that the median of all median autosomal signals is zero.~\citep{Affymetrix_2008m}

%% From the SNPs and CN probes in the default CDF, GTC/CN5 filters out an additional set of 490 SNPs that are on ChrY (or without annotation) as well as 30,033 CN probes that was found to overlap ($\pm10$ base pairs from the center base) with known SNPs in dbSNP.

\textit{Important limitations}:
It is not possible in current implementation of CN5 to estimate raw CN ratios on ChrX using ``males'' as well.  Moreover, since the software does not export $\{\theta_{ij}\}$ either, it is not possible solve this problem by calculating the log-ratios externally.  In order to do a fair evaluation, we instead have to calculate ChrX CNs for the other methods based on ``females'' only.

The CN5 method is only available for \GWSSix.  Affymetrix explicitly says that neither GTC nor APT supports it for \GWSFive.

It is also limited to the default CDF, that is, it cannot be used with \filename{GenomeWideSNP\_6.Full.cdf} (this is because the implemented model is 'hardwired' to the default CDF).



\subsection*{dChip}
For the dChip model, we used the \pkg{dChip~2008} (Build: July 10, 2008, \url{http://www.dchip.org/}).  Probe-level data was normalized using the \emph{invariant-set method}~\citep{LiWong_2001}, and PM signals were background corrected by '5th percentile of region (PM-only)'.  Array 'NA12750' was suggested by dChip to be used as the baseline array for normalization, because it had the median median (sic!) probe signal.  As suggested, we verified that the spatial intensity plot of this array was not abnormal.
For probe summarization, the dChip multiplicate model was used, with $\PM=\PMA+\PMB$ for SNPs (``Compute signals separately for A and B allele'' unchecked), returning MBEI scores (corresponding to $\{\theta_{ij}\}$).  For maximal comparison, the MBEI scores were imported to \pkg{aroma.affymetrix} and raw CNs where calculated as in Eqn~\eqref{eqnLogRatio}.

%% We used GenomeWideSNP_6.Full.cdf
%% [GenomeWideSNP_6.Full.cdf.bin: 220,277,412 bytes]
%% To load all 60 arrays into memory in dChip takes roughly 1.4GB.  /HB 2008-08-19
%% In order to do PLM in dChip for GWS6, one has to set the memory to smallest possible, i.e. 250MB, which will make dChip allocate approx 1.6GB.  With 500MB (or 1000MB), dChip will try to allocate more than 2GB and because Windows won't give more than 2GB to a process, you will get an ``Out of Memory'' error dialog.  There is a ``/3GB'' Windows option I haven't considered. /HB 2008-08-19

\subsection*{dChip*}
Due to odd performance of dChip for SNPs, we also ran the analysis where the MBEI probe summarization was replaced by averaging the signals while keeping everything else the same.  We denote this flavor of the dChip method by adding an asterisk to the label.


%% We are not aware of other freely available tools that can easily process \GWSFive or \GWSSix data for copy number analysis.

\begin{table}
\begin{center}
\begin{tabular}{lcccc}
                             &  APT      & CRMA6     & dChip     & GTC~v2  \\
\hline
GenomeWideSNP\_5 (default)   &  yes      & yes       & yes       & no      \\
GenomeWideSNP\_5 (full)      &  yes      & yes       & yes       & no      \\
GenomeWideSNP\_6 (default)   &  yes      & yes       & yes       & yes     \\
GenomeWideSNP\_6 (full)      &  yes      & yes       & yes       & no      \\
\hline
\end{tabular}
\end{center}
\caption{Summary of methods for estimating raw CNs.}
\label{tblSummaryOfMethods}
\end{table}



\subsection*{Methods for the evaluation}
We base all the performance assessments using relative copy numbers (chip effects) on the non-logaritmic scale, that is, $C_{ij}=2\cdot\theta_{ij}/\theta_{Rj}$.  This contrary to \cite{BengtssonH_etal_2008a}, where we used log-ratios $M_{ij}=\log_2(\theta_{ij}/\theta_{Rj})$.  
%% If there are no negative $C_{ij}$, the full-resolution ROC estimates are identical, because the ROC is invariant under strictly increasing transforms.

Smoothing...


\subsection*{Sets of loci used for the evaluation}

We use ChrX and ChrY loci for the evaluation with pseudo-autosomal regions (PARs) and known CN polymorphic (CNPs) regions excluded.  
According to Affymetrix' NetAffx annotation release 26 there are 87,204 ChrX and 9,486 ChrY loci in \GWSSixf.Full.CDF.

The PARs for ChrX are PAR1=$[1,2692881]$ and PAR2=$[154494747,154824264]$, and for ChrY they are PAR1=$[1,2692881]$ and PAR2=$[57372174,57701691]$~\citet{BlaschkeRappold_2006}.

For CNP regions, we use the ones identified by \citet{RedonR_etal_2006}.  There are 48 such regions on ChrX and and 7 on ChrY.  We use a safetly margin of 100kb on each side.




%% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% % Assessment of excluded loci
%% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% \section*{Additional results}
%% 
%% \subsection*{Assessment of excluded loci}
%% 
%% Since the full CDF contains additional SNPs that could potentially be used to estimate raw CNs, we wanted to investigate whether these loci can be used to identify single copies from diploids.  For GWS6, there are 1,115 (1.30\%) SNPs excluded from ChrX, which is representative to the amount excluded from an average chromosome (1.35\%) (Section 'Full and filtered sets of loci').
%% %% GWS5: 644 (2.39\%) ChrX SNPs excluded.
%% When filtering out loci in CN polymorphic and pseudo-autosomal regions, there are 919 ChrX loci left for evaluation on GWS6.  The ROC curves for the default and the excluded set of loci are shown in Figure~\ref{figROC-excluded}.  Although the excluded loci have lower TP rates, it is still clear that they have discriminatory power.
%% \begin{figure}[!tpbh]
%% \begin{center}
%%  \resizebox{0.38\columnwidth}{!}{\includegraphics{HapMap270,6_0,CEU,founders,plotROC,excludedLoci,all,y0=1_00-col}}%
%% \end{center}
%%  \caption{
%%   Performance of an average ChrX locus that has been excluded (919 loci) from the default (\GWSSix) CDF compared with the average ChrX locus in the default CDF (75,642 loci).  The ROC curves are based on raw-CN estimates from CRMA6.
%%  }
%%  \label{figROC-excluded}
%% \end{figure} 
%% 
%% This suggest that those loci could be included in downstream analysis in order to increase the effective resolution.  However, when identifying CN aberrations using segmentation methods, the assumption is that the raw CNs within a region have the same means.  When comparing density estimates of the raw CNs from the excluded loci with a same-size random set of default loci, we find, contrary to this general assumption, that there is a strong bias in the excluded loci.  This is likely the reason why Affymetrix excluded them (ref?).  The diploid loci have the same mean (0.00) and approximately the same standard deviation (0.20 v. 0.18), whereas the single-copy loci have a significantly higher mean for the excluded (-0.35) compared with the default (-0.65) set of loci.  This strong bias in the mean is likely to cause problems for a segmentation method.  
%% When calculating the TP rate as function of resolution using the method in \cite{BengtssonH_etal_2008a}, which also relies on the assumption that neighboring loci should have the same mean level, the effective resolution does indeed decrease slightly when we include the excluded loci (not shown).
%% Furthermore, the excluded single-copy loci have a greater standard deviation (0.37) compared with the default loci (0.28) which explains the differences in ROC performances in Figure~\ref{figROC-excluded}.
%% \begin{figure}[!tpbh]
%% \begin{center}
%%  \resizebox{0.38\columnwidth}{!}{\includegraphics{HapMap270,6_0,CEU,founders,plotRawCNsDensity,excluded,all-col}}%
%%  \resizebox{0.38\columnwidth}{!}{\includegraphics{HapMap270,6_0,CEU,founders,plotRawCNsDensity,default,all-col}}%
%% \end{center}
%%  \caption{
%%   Empirical densities for the 30 males and 29 females of the raw CNs for the 919 excluded loci (left) and 919 random loci from the default \GWSSix CDF (right).  The diploid distributions are approximately the same for both sets, but for the single-copy loci the mean level (-0.35) of the excluded loci is substantially different from the default loci (-0.66).  The standard deviation is 0.37 and 0.26, respectively.
%%  }
%%  \label{figRawCN-excluded}
%% \end{figure}
%% Because we do not know the bias of these loci at other copy number levels, but also because we do not know their biases on the other chromosomes, it is not obvious how to correct for the biases.  An alternative to filtering out loci is to downweight them, according to some weight function, when fitting the segmentation models.  However, we do not know of any segmentation methods that take weights.  
%% 
%% 
%% \subsection*{Assessment of GWS5 estimates}
%% In this section we give a brief summary of the results from estimating raw CNs in the \GWSFive assay using the CRMA6, the APT, and the dChip method.  The GTC software does not support the \GWSFive assay.
%% 
%% [Will run this when Affymetrix release the corrected annotation data...]


\end{document}
